### PR TITLE
Support Ukraine and Moldova in the country list and in the dataset.

### PR DESCRIPTION
**Pull request by euronion**
## TODO list

New country keys to be supported:
- `MD`: [Moldova, Republic of](https://en.wikipedia.org/wiki/Moldova)
- `UA`: [Ukraine](https://en.wikipedia.org/wiki/Ukraine)

TODOs

- [x] Crimea: Is automatically included or not when using "UA" key (`country_shapes.json`)? (Crimea not included
- [x] Keep Crimea as part of "UA" key or not?
- [x] Updated power plants list incl. Moldova @FabianHofmann 
- [x] OPSD (load demand) @martacki 
- [x] ENTSO-E grid extract
- [x] 750 kV transmission electrical params @fneum @p-glaum 
- [x] 750 kV lines types -> point release PyPSA to use with pypsa-eur @fneum -> 0.19.2 (conda may take 24h)
- [x] 750 kV transmission include in simplify_network & enable in pypsa-eur config line type & enforce min. version in `environment.yaml` for pypsa point release @p-glaum (alt.: 380kV assumptions)
- [x] GDP/POP data for assigning load data to bus regions. @euronion @martacki (decided to use GDP data only, i.e. no population)
- [x] No pop data considered for `add_electricity` for MD, UA (simlification)
- [x] Cutout containing Ukraine: ERA5 @euronion @fneum 
- [x] Cutout containing Ukraine: SARAH @euronion
- [x] Hydro profiles for Ukraine (Moldova has only ~100 MW hydro if seen correctly, so less relevant) @euronion [Needs update to databundle, see below]
- [x] Renewable potentials?

## Additions to databundle

- [x] EIA data (https://github.com/PyPSA/pypsa-eur/pull/321#issuecomment-1059728440)
- [x] @martacki gdp data for UA, MD (Link: https://github.com/martacki/pypsa-ua_gdp_data)
- [x] Update cutouts from [here](https://fex.hrz.uni-giessen.de/fop/YeKwKF9F/cutouts.tar) @fneum https://zenodo.org/record/6350001#.Yi3p9Lko8UE

## Release

- [x] Release notes

## Checklist

- [ ] I tested my contribution locally and it seems to work fine.
- [ ] Code and workflow changes are sufficiently documented.
- [ ] Newly introduced dependencies are added to `envs/environment.yaml` and `envs/environment.docs.yaml`.
- [ ] Changes in configuration options are added in all of `config.default.yaml`, `config.tutorial.yaml`, and `test/config.test1.yaml`.
- [ ] Changes in configuration options are also documented in `doc/configtables/*.csv` and line references are adjusted in `doc/configuration.rst` and `doc/tutorial.rst`.
- [ ] A note for the release notes `doc/release_notes.rst` is amended in the format of previous release notes.

### CODE REVIEW

**Comment by fneum** [comment]
So one problem currently is that PyPSA-Eur thinks UA is synchronized with continental Europe, which it is not. Well not all of it, there is a small Island in the border region that is synchronized with continental Europe. I tried to find out where the B2B converters are, but I couldn't find them. As all studies EU + UA I can think of will assume synchronisation, I'd propose to keep synchronised.

If somebody wants to extend this investigation: 

- https://www.lowcarbonukraine.com/wp-content/uploads/Synchronising-Ukraine%C2%B4s-and-Europe%C2%B4s-electricity-grids.pdf
- https://en.wikipedia.org/wiki/Burshtyn_TES
- https://en.wikipedia.org/wiki/Rzesz%C3%B3w%E2%80%93Khmelnytskyi_powerline
- https://en.wikipedia.org/wiki/Vetrino%E2%80%93Isaccea%E2%80%93Yuzhnoukrainsk_powerline
- https://en.wikipedia.org/wiki/Albertirsa%E2%80%93Zakhidnoukrainska%E2%80%93Vinnytsia_powerline

File: scripts/add_electricity.py [hunk 1]
```diff
@@ -90,6 +90,7 @@
 import pandas as pd
 import numpy as np
 import xarray as xr
+import xagg as xa
```
Comment (euronion) [inline scripts/add_electricity.py:93]: TODO env

File: scripts/add_electricity.py [hunk 2]
```diff
@@ -189,21 +190,34 @@ def load_powerplants(ppl_fn):
...
                .reindex(substation_lv_i))
     opsd_load = (pd.read_csv(load, index_col=0, parse_dates=True)
                 .filter(items=countries))
 
+    #ToDo: adapt time+slices from config etc. (cover all data)
+    gdp = (xr.open_dataset(gdp)
+           .sel(time=2015)
+           .sel(longitude=slice(10,30))
+           .sel(latitude=slice(50, 30)))
+    weightmap = xa.pixel_overlaps(gdp, regions.iloc[0:2])
```
Comment (euronion) [inline scripts/add_electricity.py:206]: TODO for all regions

File: scripts/add_electricity.py [hunk 3]
```diff
@@ -189,21 +190,34 @@ def load_powerplants(ppl_fn):
...
+def attach_load(n, regions, load, nuts3_shapes, gdp, countries, scaling=1.):
 
     substation_lv_i = n.buses.index[n.buses['substation_lv']]
     regions = (gpd.read_file(regions).set_index('name')
                .reindex(substation_lv_i))
     opsd_load = (pd.read_csv(load, index_col=0, parse_dates=True)
                 .filter(items=countries))
 
+    #ToDo: adapt time+slices from config etc. (cover all data)
+    gdp = (xr.open_dataset(gdp)
```
Comment (euronion) [inline scripts/add_electricity.py:202]: TODO full region, not only slice

File: scripts/add_electricity.py [hunk 4]
```diff
@@ -189,21 +190,34 @@ def load_powerplants(ppl_fn):
...
+           .sel(time=2015)
+           .sel(longitude=slice(10,30))
+           .sel(latitude=slice(50, 30)))
+    weightmap = xa.pixel_overlaps(gdp, regions.iloc[0:2])
+    aggregated = xa.aggregate(gdp, weightmap)
+
+    print(aggregated.to_dataset().name)
```
Comment (euronion) [inline scripts/add_electricity.py:209]: TODO should be list of bus IDs, or find correct one

File: scripts/add_electricity.py [hunk 5]
```diff
@@ -218,12 +232,18 @@ def upsample(cntry, group):
             # relative factors 0.6 and 0.4 have been determined from a linear
             # regression on the country to continent load data
             factors = normed(0.6 * normed(gdp_n) + 0.4 * normed(pop_n))
+            if cntry in ['UA', 'MD']:
+                #generate new factors in this case
+                print('ToDo: adjust load for UA and MD here')
```
Comment (euronion) [inline scripts/add_electricity.py:237]: TODO factor based on gdp / pop dens for UA, MD

**Comment by euronion** [comment]
`build_hydro` requires input from the EIA. The last time we retrieved the data from there was 2017, they changed their data access and csv format since then quite abit. For prototyping, instead of adjusting all scripts and rule, I've added the values for Ukraine and Moldova to the EIA data file. See attached.

[EIA_hydro_generation_2000_2014.csv](https://github.com/PyPSA/pypsa-eur/files/8190363/EIA_hydro_generation_2000_2014.csv)

- [ ] Needs to be added to the databundle.

**Comment by martacki** [comment]
if you want to run the last commits yourself (`add_electricity`), you need to generate or download the input data `GDP_PPP_30arcsec_v3_mapped.csv` once and move the file into `pypsa-eur/data/bundle`.

Available here:
https://github.com/martacki/pypsa-ua_gdp_data

(Caution: _Quick&Dirty solution, runs very slow!_)

**Comment by euronion** [comment]
> if you want to run the last commits yourself (`add_electricity`), you need to generate or download the input data `GDP_PPP_30arcsec_v3_mapped.csv` once and move the file into `pypsa-eur/data/bundle`. **edit**: in case you download the data and don't re-generate it, you should rename `GDP_PPP_30arcsec_v3_mapped_default.csv` to `GDP_PPP_30arcsec_v3_mapped.csv`
> 
> Available here: https://github.com/martacki/pypsa-ua_gdp_data
> 
> Caution: _Quick&Dirty solution; in case you execute the script - it runs very slow!! (~2h on my local machine)_

How does the quality of distributing load based on this look? We two already talked about it, but we could use the same datasets and roll-out the same approach for everywhere and instead drop the NUTS3 approach currently used.

**Comment by fneum** [comment]
PyPSA 0.19.2 is out on PyPI with 750kV line type, conda may take 24h.

**Comment by fneum** [comment]
new cutouts now available at: https://zenodo.org/record/6350001#.Yi3p9Lko8UE

**Comment by euronion** [comment]
Looking at the potentials for renewables for UA, MD, @fneum was right: The potentials are currently 0, as Corine is not extending to there.

I quickly used the Copernicus Land Cover dataset to see how well that could substitute Corine, here the results (onshore wind only):

image

The match is pretty poor for densily populated areas. I believe that is because the Copernicus dataset does not differentiate between build-on areas, while the Corine dataset differentiates between different types of buildings.

**Comment by euronion** [comment]
We could set the renewable potentials to 0 (current status), set them to `np.inf` (hacky, but ok for now) or manually set them to values based on the Copernicus dataset (quick and dirty, hardcoded, probably the most accurate).

What do you think?

**Comment by fneum** [comment]
+1 for hardcoded Copernicus for UA/MD
profiles look ok?

**Comment by coroa** [comment]
Just a shout-out that in principle one could also switch to using the ESA-CCI land cover dataset which has global coverage: https://www.esa-landcover-cci.org/

**Comment by energyLS** [comment]
@fneum do you still need someone to test run this on a fresh pypsa-eur clone? Could do it this week

**Comment by fneum** [comment]
would be great! Thanks @energyLS !

**Comment by energyLS** [comment]
Successful. Steps I have done:
Fresh repo: `git clone https://github.com/PyPSA/pypsa-eur.git`
Switch to branch [ukraine_hackathon](https://github.com/PyPSA/pypsa-eur/tree/ukraine_hackathon)
Changed config to `clusters: [39, 128]` and `opts: [Co2L-144H]` to reduce solving time
Run: `snakemake -c1 solve_all_networks`

Worked all fine, did some additional plotting of the solved network and looks good.
Ready to merge from my point of view @fneum

**Comment by pz-max** [comment]
Just for reference. This PR is merged because of ENTSO-E news:

_"Continental European TSOs announce completion of synchronisation project with Ukrenergo and significant increase in export capacity from Continental Europe to Ukraine. https://www.entsoe.eu/news/2023/11/28/continental-european-tsos-announce-completion-of-synchronisation-project-with-ukrenergo-and-significant-increase-in-export-capacity-from-continental-europe-to-ukraine/"_ :eu:

**Pull request merged**